### ISSUE TITLE
Docs for require-default-props are misleading

### ISSUE THREAD

**Issue by aij** [yannickcr/eslint-plugin-react#1498]
> One advantage of defaultProps over custom default logic in your code is that defaultProps are resolved by React before the PropTypes typechecking happens, so typechecking will also apply to your defaultProps.

Of couse, `require-default-props` only applies where the prop type is optional, in which case providing a default via `defaultProps` is insufficient. If the component can't *also* handle `null` and `undefined` then the prop type should instead be marked as `.isRequired` so that an appropriate proptype warning can be issued if `null` or `undefined` are passed in place of the default prop.

**Comment by oliverjam** [comment]
Yeah we've run into this a bit using Flow. If a prop is typed as a Maybe it could be null _or_ undefined. Adding a default prop doesn't help here because a null value prop won't be replaced by the default.

e.g.

```jsx
type Props = {
  quantities?: Array<number>
}

const Thing = ({ quantities }: Props) => (
  {quantities.map(quantity => <li>...</li>)}
)

Thing.defaultProps = {
  quantities: [0]
}
```

won't trigger this rule but will break at runtime if we pass `quantities` as `null`

**Comment by aij** [comment]
@oliverjam It's not just `null`. An `undefined` value prop *also* won't be replaced by the default, which can be especially confusing to talk about. The default props are only used when the prop is absent.

For your example `Thing` component, `quantities` should be marked required, because it is required, and rendering it as `<Thing/>` is still fine because the required prop is being supplied via `defaultProps`.

**Comment by ljharb** [comment]
@aij that’s incorrect; `undefined` will always be replaced by a defaultProp or an argument default. React doesn’t differentiate absent from `undefined` (which is the proper thing to do)

**Comment by aij** [comment]
@ljharb Huh, I've definitely been bitten by that in the past, but trying it again now defaultProps are being used even when `undefined` is explicitly passed.

I wonder if that changed at some point (possibly since the React 0.14 days), or if we had some weird tooling issue.

Anyway, good to know React does the proper thing (now). :smile:

**Comment by ljharb** [comment]
Since 0.13 at least, that's how it's behaved; I can't speak to before that.

**Comment by oliverjam** [comment]
Yeah the prop type has to be a Maybe because it _might_ be undefined. This value is coming from our GraphQL API, and GraphQL uses null for missing values. So the only real way for us to handle this is to make the type optional and let Flow force us to handle the potential null value inside the component.

Setting a default prop doesn't help at all in that situation, which is why I think the rule is a little pointless

**Comment by ljharb** [comment]
@oliverjam if you're suggesting a modification to the rule (like an option) that would make it more useful when using flow, i'd love to hear about it! if you're just suggesting that the rule doesn't make sense at all when using flow, then you can just turn it off :-)
